### PR TITLE
Refactor channel tests.

### PR DESCRIPTION
**Pull request by beauby**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #216 Add support for CUDA channels in the Pipe.
* **#214 Refactor channel tests.**
* #213 Get rid of `channelName()` in channel test helper.
* #212 Make Channel API accept buffer structs rather than raw pointers.
* #211 Add TENSORPIPE_SUPPORTS_CUDA flag.
* #210 Move channel types from `channel::Channel::` to `channel::`.

Make generic channel tests work with either CPU or CUDA channels:
+ Move from defining tests within the `TEST_P` macro to defining
standalone test classes. This is needed because `TEST_P` does not play
well with templated fixture classes. While it has the downside of
requiring an extra step to register a test after it has been
defined (by calling `CHANNEL_TEST_GENERIC(Foo);`), it improves
readability of tests by having them override `server()`/`client()`
methods rather than call `testConnection()` with two lambdas.
+ Data wrapping (for channel-type agnostic tests) is now done through
a templated `DataWrapper` class rather than through a virtual method
on the `ChannelTestHelper`.

Differential Revision: [D23598040](https://our.internmc.facebook.com/intern/diff/D23598040)

### CODE REVIEW

File: tensorpipe/test/channel/basic/basic_test.cc [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 namespace {
 
-class BasicChannelTestHelper : public ChannelTestHelper {
+class BasicChannelTestHelper : public ChannelTestHelper<tensorpipe::CpuTensor> {
```
Comment (lw) [inline tensorpipe/test/channel/basic/basic_test.cc:14]: Wondering if we should introduce some `CpuChannelTestHelper` shortcuts?

File: tensorpipe/test/channel/channel_test.h [hunk 2]
```diff
@@ -20,69 +20,156 @@
...
+  return future;
+}
+
+template <typename TTensor>
+class ChannelTest {
  public:
-  ChannelTest() : helper_(GetParam()), peers_(helper_->makePeerGroup()) {}
+  virtual void run(ChannelTestHelper<TTensor>* helper) {
+    helper_ = helper;
+    peers_ = helper_->makePeerGroup();
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:171]: Doing this in the run method means that any test case that does _not_ want to use the server/client method patter must re-do this, which doesn't scale well.

Also, the fact that the default run method "forces" the server/client pattern may be odd, as users then have to override it and make it do an entirely different thing.

What we could do instead is:
- give ChannelTest an "entry point" method, which is the one called from the `TEST_P` macros below.
- give ChannelTest a pure virtual "run" method, which is called by the above entry-point method.
- have a ClientServerChannelTest subclass, which implements that run method, and in them calls the server/client methods, which are defined as purely virtual.

This way someone who wants to do a non-server/client test can subclass ChannelTest, avoid having to set up the helper_ and peers_ fields, and then implement run (without overriding an already-existing implementation).

Note that if the above suggestion (of using the constructor) works then the entry point method wouldn't be doing anything, and we could simplify it and have TEST_P directly call run.

File: tensorpipe/test/channel/channel_test.cc [hunk 3]
```diff
@@ -15,513 +15,568 @@
...
-      });
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Initialize with sequential values.
+    std::vector<uint8_t> data(dataSize);
+    std::iota(data.begin(), data.end(), 0);
+    DataWrapper<TTensor> wrappedData(data);
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:45]: What does creating a `DataWrapper` directly buy us, instead of using the helper's `makeBuffer` method which you removed?

File: tensorpipe/test/channel/channel_test.cc [hunk 4]
```diff
@@ -15,513 +15,568 @@
...
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, wrappedData.tensor());
+    Error descriptorError;
+    TDescriptor descriptor;
+    std::tie(descriptorError, descriptor) = descriptorFuture.get();
+    EXPECT_FALSE(descriptorError) << descriptorError.what();
+    this->peers_->send(PeerGroup::kClient, descriptor);
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:56]: This sounds like a C++ thing I'm unaware of, but why do we suddenly need `this->`?

File: tensorpipe/test/channel/channel_test.cc [hunk 5]
```diff
@@ -15,513 +15,568 @@
...
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
 };
 
-TEST_P(ChannelTest, ServerToClient) {
+CHANNEL_TEST_GENERIC(ClientToServer);
+
+// TEST_P(ChannelTest, ServerToClient) {
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:95]: What's this?

File: tensorpipe/test/channel/channel_test.cc [hunk 6]
```diff
@@ -15,513 +15,568 @@
...
+
+    ctx->join();
+  }
+};
+
+CHANNEL_TEST_GENERIC(ServerToClient);
+
+template <typename TTensor>
+class SendMultipleTensorsTest : public ChannelTest<TTensor> {
+  const int dataSize = 256 * 1024; // 256KB
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:161]: This was a `constexpr`, why the change?

File: tensorpipe/test/channel/channel_test.cc [hunk 7]
```diff
@@ -15,513 +15,568 @@
...
+
+    ctx->join();
+  }
+
+  void client(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("client");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
+
+    std::vector<std::unique_ptr<DataWrapper<TTensor>>> wrappedDataVec;
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:207]: Do we need to wrap the DataWrappers in unique_ptrs because they are not movable? Would it make sense perhaps to make them movable? As it stands, we're making plenty of heap allocations just so they can hold a few pointers...

File: tensorpipe/test/channel/channel_test.cc [hunk 8]
```diff
@@ -15,513 +15,568 @@
...
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) = sendWithFuture(
+        channel,
+        TTensor{
+            .ptr = nullptr,
+            .length = 0,
+        });
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:375]: This is not generic. We should either find a way to make it generic or, as there is no guarantee (in principle) that for every possible device type its data can be represented as a pointer, we shouldn't make this test generic but have an instance of Cpu and (if we so choose) one for Cuda.

File: tensorpipe/test/channel/channel_test.cc [hunk 9]
```diff
@@ -15,513 +15,568 @@
...
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TTensor>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Allocate a non-empty vector so that its .data() pointer is non-null.
+    std::vector<uint8_t> data(1);
+    DataWrapper<TTensor> wrappedData(data);
+    TTensor tensor = wrappedData.tensor();
+    tensor.length = 0;
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:426]: Similarly non-generic.

File: tensorpipe/test/channel/channel_test.cc [hunk 10]
```diff
@@ -15,513 +15,568 @@
...
+    // Perform send and wait for completion.
+    std::promise<std::tuple<Error, TDescriptor>> descriptorPromise;
+    std::promise<Error> sendPromise;
+    std::mutex mutex;
+    std::unique_lock<std::mutex> callerLock(mutex);
+    channel->send(
+        CpuTensor{
+            .ptr = data.data(),
+            .length = data.size(),
+        },
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:518]: Goes back to a comment I left on an earlier diff: why use "keyword" args instead of "positional" args? It seems we're mixing the two styles.

File: tensorpipe/test/channel/channel_test.cc [hunk 11]
```diff
@@ -15,513 +15,568 @@
...
+        [&recvPromise, &mutex](const Error& error) {
+          std::unique_lock<std::mutex> calleeLock(mutex);
+          recvPromise.set_value(error);
+        });
+    callerLock.unlock();
+    Error recvError = recvPromise.get_future().get();
+    EXPECT_FALSE(recvError) << recvError.what();
+
+    // Validate contents of vector.
+    for (auto i = 0; i < dataSize; i++) {
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:571]: Also a reference to a previous comment: can we stick to dataSize or data.size() consistently?

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc [hunk 12]
```diff
@@ -17,70 +17,23 @@
 
 namespace {
 
-class CudaWrapper : public DataWrapper {
+class CudaChannelTestHelper : public ChannelTestHelper<tensorpipe::CudaTensor> {
```
Comment (lw) [inline tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:20]: Shouldn't this be called `CudaIpcChannelTestHelper`? (Note the `Ipc`)

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc [hunk 13]
```diff
@@ -17,70 +17,23 @@
...
-
-//   std::shared_ptr<DataWrapper> makeBuffer(size_t len) override {
-//     return std::make_shared<CudaWrapper>(len);
-//   }
-// };
-
-// CudaChannelTestHelper helper;
-
-class CudaIpcChannelTest : public ChannelTest {};
+class CudaIpcChannelTest : public CudaChannelTest {};
```
Comment (lw) [inline tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:36]: What is this doing exactly? I don't see anything equivalent in the other channels.

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc [hunk 14]
```diff
@@ -92,113 +45,115 @@ using namespace tensorpipe::channel;
...
+    TP_CUDA_CHECK(cudaFree(ptr));
+
+    this->peers_->done(PeerGroup::kClient);
+    this->peers_->join(PeerGroup::kClient);
+
+    ctx->join();
+  }
+};
+
+CHANNEL_TEST(CudaIpc, ReceiverWaitsForStartEvent);
```
Comment (lw) [inline tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:153]: I thought the first argument of `CHANNEL_TEST` is the type of the tensor, which should be `Cuda` here. In fact, this test (`ReceiverWaitsForStartEvent`) should probably go in a CUDA test suite, and then this CUDA test suite should be instantiated once for CudaIpc. Right?

**Review by lw — COMMENTED** [review]
I feel like this diff is doing two things at once: create two hierarchies of ChannelTest/ChannelTestHelper by templating them, and moving from the testConnection method to the server/client virtual methods (plus a bunch of other small things). It would have been clearer if they had been done in separate diffs.

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc (hunk 14 above)
Comment (beauby) [reply tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:153]: The first argument of `CHANNEL_TEST` is a prefix for a test suite, for instance `Foo` -> `FooChannelTest`. It coincides with the type of the tensor in the case of generic test suites, but that's just a convenience.

File: tensorpipe/test/channel/channel_test.cc (hunk 3 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc]: With the current setting, the `makeBuffer()` method would simply forward its parameter to the constructor of `DataWrapper` and return that, it seems like an unnecessary indirection.

File: tensorpipe/test/channel/channel_test.cc (hunk 5 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc]: Removing it.

File: tensorpipe/test/channel/channel_test.cc (hunk 7 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc]: Yes, making them movable is probably the right thing to do here. Will update.

File: tensorpipe/test/channel/channel_test.cc (hunk 8 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc]: Sounds good.

File: tensorpipe/test/channel/channel_test.cc (hunk 10 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.cc]: Will fix.

File: tensorpipe/test/channel/channel_test.h [hunk 15]
```diff
@@ -20,69 +20,156 @@
...
-  std::shared_ptr<PeerGroup> peers_;
-
+template <typename TTensor>
+[[nodiscard]] std::pair<
+    std::future<
+        std::tuple<tensorpipe::Error, tensorpipe::channel::TDescriptor>>,
+    std::future<tensorpipe::Error>>
+sendWithFuture(
+    std::shared_ptr<tensorpipe::channel::Channel<TTensor>> channel,
+    TTensor tensor) {
```
[the remark this answers is not in the document]
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h]: As we moved to pass by copy for tensors, let's just keep it non const.

File: tensorpipe/test/channel/channel_test.h [hunk 16]
```diff
@@ -20,69 +20,156 @@
...
+        promise->set_value(error);
+      });
+  return future;
+}
+
+template <typename TTensor>
+class ChannelTest {
  public:
-  ChannelTest() : helper_(GetParam()), peers_(helper_->makePeerGroup()) {}
+  virtual void run(ChannelTestHelper<TTensor>* helper) {
```
[the remark this answers is not in the document]
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h]: Inheriting from `TestWithParams` here won't work, but I have an idea to avoid having to set those in `run()`. Will update the PR.

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc (hunk 12 above)
Comment (beauby) [reply tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc]: Good catch, will update.

File: tensorpipe/test/channel/basic/basic_test.cc (hunk 1 above)
Comment (beauby) [reply tensorpipe/test/channel/basic/basic_test.cc]: Will do.

File: tensorpipe/test/channel/channel_test.cc (hunk 4 above)
Comment (lw) [reply tensorpipe/test/channel/channel_test.cc:56]: Damned C++, this is ADL all over again. I'm thankful I've learned about this the gentle way though...

File: tensorpipe/test/channel/channel_test.cc (hunk 6 above)
Comment (lw) [reply tensorpipe/test/channel/channel_test.cc:160]: Ok, if it doesn't work don't bother (but maybe add a comment so that if anyone who comes around knows they can help).

File: tensorpipe/test/channel/channel_test.cc (hunk 9 above)
Comment (lw) [reply tensorpipe/test/channel/channel_test.cc:418]: It looks like one comment I though I had made before has disappeared, I'm rewriting it now.

File: tensorpipe/test/channel/channel_test.cc [hunk 17]
```diff
@@ -15,495 +15,556 @@
...
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TBuffer>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, TBuffer{nullptr, 0});
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.cc:370]: This is not a "generic" test, because it assumes that TBuffer can be constructed using two arguments: a pointer and a length. This may be the case with Cpu and CudaBuffer, but it's not guaranteed to always be the case. This test aims to test whether a channel can handle a null pointer but, in principle, a TBuffer may not even hold a pointer. So maybe it's better to make this become a Cpu-specific test? (And possibly create a copy of it for Cuda channels, if we decide we want Cuda channels to also be able to handle null pointers)

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc (hunk 14 above)
Comment (lw) [reply tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:153]: Right, but (also relating to the comment above), this test isn't specific to the `cuda_ipc` channel, it could be run on all Cuda channels (and should work for all of them). So shouldn't we add it to the "generic" Cuda test suite, rather than to the one specific to this one channel?

File: tensorpipe/test/channel/channel_test.h [hunk 18]
```diff
@@ -20,69 +20,156 @@
...
+      std::move(descriptor),
+      std::move(tensor),
+      [promise{std::move(promise)}](const tensorpipe::Error& error) {
+        promise->set_value(error);
+      });
+  return future;
+}
+
+template <typename TTensor>
+class ChannelTest {
```
[the remark this answers is not in the document]
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h]: Good idea. I separated the base `ChannelTestCase<>` from its subclass `ClientServerChannelTestCase<>`, and renamed the test suites `CpuChannelTestSuite/CudaChannelTestSuite`.

File: tensorpipe/test/channel/channel_test.h (hunk 2 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h]: The problem with using a constructor (with a parameter) is that it won't be inherited. The current trade off I came up with is as follows: the helper is passed as a parameter to `run()` in every case, and when, for a specific pattern, the helper needs to be made available from other methods (which currently only happens in the ClientServerChannelTestCase), this is done in `run()`.

File: tensorpipe/test/channel/channel_test_cpu.cc [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+CHANNEL_TEST(CpuChannelTestSuite, EmptyTensor);
+
+// This test wants to make sure that the "heavy lifting" of copying data isn't
+// performed inline inside the recv method as that would make the user-facing
+// read method of the pipe blocking.
+// However, since we can't really check that behavior, we'll check a highly
+// correlated one: that the recv callback isn't called inline from within the
+// recv method. We do so by having that behavior cause a deadlock.
+class CallbacksAreDeferredTest
+    : public ClientServerChannelTestCase<tensorpipe::CpuBuffer> {
```
Comment (lw) [inline tensorpipe/test/channel/channel_test_cpu.cc:121]: The tests above do not need the namespace qualifier
```suggestion
    : public ClientServerChannelTestCase<CpuBuffer> {
```

File: tensorpipe/test/channel/channel_test_cuda.cc [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+#include <cuda_runtime.h>
+
+#include <tensorpipe/test/channel/channel_test.h>
+#include <tensorpipe/test/channel/kernel.cuh>
+
+using namespace tensorpipe;
+using namespace tensorpipe::channel;
+
+class ReceiverWaitsForStartEventTest
+    : public ClientServerChannelTestCase<tensorpipe::CudaBuffer> {
```
Comment (lw) [inline tensorpipe/test/channel/channel_test_cuda.cc:18]: Also probably not needed?
```suggestion
    : public ClientServerChannelTestCase<CudaBuffer> {
```

File: tensorpipe/test/channel/channel_test.h [hunk 21]
```diff
@@ -73,6 +70,14 @@ class DataWrapper<tensorpipe::CudaBuffer> {
     }
   }
 
+  explicit DataWrapper(DataWrapper&& other)
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:73]: Nit: doesn't the "rule of 5" say that whenever one defines a move constructor one should also define a move assignment operator?

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc [hunk 22]
```diff
@@ -31,129 +30,17 @@ class CudaChannelTestHelper : public ChannelTestHelper<tensorpipe::CudaBuffer> {
   }
 };
 
-CudaChannelTestHelper helper;
+CudaIpcChannelTestHelper helper;
 
-class CudaIpcChannelTest : public CudaChannelTest {};
+class CudaIpcChannelTestSuite : public CudaChannelTestSuite {};
```
Comment (lw) [inline tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc:35]: Now there are no IPC-specific CUDA tests anymore. Do we still need this?

File: tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc [hunk 23]
```diff
@@ -31,129 +30,17 @@ class CudaChannelTestHelper : public ChannelTestHelper<tensorpipe::CudaBuffer> {
...
-INSTANTIATE_TEST_CASE_P(CudaIpc, CudaChannelTest, ::testing::Values(&helper));
 INSTANTIATE_TEST_CASE_P(
     CudaIpc,
-    CudaIpcChannelTest,
+    CudaChannelTestSuite,
+    ::testing::Values(&helper));
+INSTANTIATE_TEST_CASE_P(
+    CudaIpc,
+    CudaIpcChannelTestSuite,
     ::testing::Values(&helper));
```
Comment (lw) [inline tensorpipe/test/channel/cuda_ipc/cuda_ipc_test.cc]: Same: do we still need this?

File: tensorpipe/test/channel/channel_test.h [hunk 24]
```diff
@@ -20,69 +20,156 @@
...
+class DataWrapper<tensorpipe::CpuTensor> {
  public:
-  virtual ~DataWrapper() = default;
-  virtual void* data() = 0;
-  virtual size_t size() = 0;
-  virtual void wrap(const void* ptr) = 0;
-  virtual void unwrap(void* ptr) = 0;
+  explicit DataWrapper(size_t length) : vector_(length) {}
+
+  explicit DataWrapper(const std::vector<uint8_t>& v) : vector_(v) {}
```
[the remark this answers is not in the document]
Comment (lw) [reply tensorpipe/test/channel/channel_test.h:31]: You marked thus as resolved but you didn't change anything...

File: tensorpipe/test/channel/channel_test.h [hunk 25]
```diff
@@ -20,69 +20,156 @@
...
+
+  tensorpipe::CpuTensor tensor() const {
+    return tensorpipe::CpuTensor{
+        .ptr = const_cast<uint8_t*>(vector_.data()),
+        .length = vector_.size(),
+    };
+  }
+
+  std::vector<uint8_t> unwrap() {
```
[the remark this answers is not in the document]
Comment (lw) [reply tensorpipe/test/channel/channel_test.h:40]: You marked thus as resolved but you didn't change anything...

File: tensorpipe/test/channel/channel_test.h [hunk 26]
```diff
@@ -20,69 +20,156 @@
...
+
+    return v;
   }
 
-  void unwrap(void* ptr) override {
-    ASSERT_EQ(data_, ptr);
+  ~DataWrapper() {
+    if (length_ > 0) {
+      EXPECT_EQ(cudaSuccess, cudaFree(cudaPtr_));
+    }
```
[the remark this answers is not in the document]
Comment (lw) [reply tensorpipe/test/channel/channel_test.h:103]: You marked thus as resolved but you didn't change anything...

File: tensorpipe/test/channel/channel_test.h (hunk 2 above)
Comment (lw) [reply tensorpipe/test/channel/channel_test.h]: Isn't there a way to have subclasses inherit constructors? Something like adding `using SuperClass;` in their body. It's not automatic (a subclass may "forget" to do so) but at least it avoids duplicating the initialization code...

File: tensorpipe/test/channel/channel_test.h [hunk 27]
```diff
@@ -20,71 +20,168 @@
...
-  virtual void wrap(const void* ptr) = 0;
-  virtual void unwrap(void* ptr) = 0;
+  explicit DataWrapper(size_t length) : vector_(length) {}
+
+  explicit DataWrapper(const std::vector<uint8_t>& v) : vector_(v) {}
+
+  tensorpipe::CpuBuffer buffer() const {
+    return tensorpipe::CpuBuffer{
+        .ptr = const_cast<uint8_t*>(vector_.data()),
+        .length = vector_.size(),
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:36]: Didn't you say that this is a C++-20 thing?

File: tensorpipe/test/channel/channel_test.h [hunk 28]
```diff
@@ -20,71 +20,168 @@
...
 
-  void wrap(const void* ptr) override {
-    data_ = const_cast<void*>(ptr);
+  explicit DataWrapper(DataWrapper&& other)
+      : cudaPtr_(other.cudaPtr_),
+        length_(other.length_),
+        stream_(other.stream_) {
+    other.cudaPtr_ = nullptr;
+    other.length_ = 0;
   }
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:79]: Once you destroy the stream in the destructor you may want to reset it to `cudaStreamDefault` here.

File: tensorpipe/test/channel/channel_test.h (hunk 27 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h:36]: This is my understanding of https://en.cppreference.com/w/cpp/language/aggregate_initialization. Basically it's a C99 thing that hasn't made it in the C++ standard before C++20, but most compilers accept it IIUC. Anyways, reverting to something officially supported.

File: tensorpipe/test/channel/channel_test.h (hunk 24 above)
[the remark this answers is not in the document]
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h:31]: My bad, some rebasing gone wrong.

File: tensorpipe/test/channel/channel_test.h (hunk 2 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h]: I considered the `using SuperClass::SuperClass` route but it seemed lesss convenient than the current solution (where subclasses only define test-related methods, being `run()` in the general case, or `server()/client()` when using the `ClientServer` pattern. Note that exposing the `helper` parameter of `run()` as a class member is not needed in all cases (currently only with ClientServer), and has more to do with how `ClientServerChannelTestCase` works than general "intialization of a test case".

File: tensorpipe/test/channel/channel_test.h [hunk 29]
```diff
@@ -76,14 +74,22 @@ class DataWrapper<tensorpipe::CudaBuffer> {
...
+    other.stream_ = cudaStreamDefault;
+  }
+
+  DataWrapper& operator=(DataWrapper&& other) {
+    cudaPtr_ = other.cudaPtr_;
+    length_ = other.length_;
+    stream_ = other.stream_;
+    other.cudaPtr_ = nullptr;
+    other.length_ = 0;
+    other.stream_ = cudaStreamDefault;
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:86]: Sorry to keep nit-picking, but this isn't a valid move assignment operator: if `this` was holding a resource, this operator is "forgetting" that resource without freeing it properly. One "trick" for a valid move constructor is to `std::swap(...)` all fields. This way, the object that we're moving from will end up destroying the resource. (Note that a similar trick also works for the move constructor, although it's not necessary there because `this` can't possibly be holding a resource yet)

File: tensorpipe/test/channel/channel_test.h [hunk 30]
```diff
@@ -76,14 +74,22 @@ class DataWrapper<tensorpipe::CudaBuffer> {
...
   tensorpipe::CudaBuffer buffer() const {
-    return tensorpipe::CudaBuffer{
-        .ptr = cudaPtr_,
-        .length = length_,
-        .stream = stream_,
-    };
+    return tensorpipe::CudaBuffer{cudaPtr_, length_, stream_};
   }
 
   std::vector<uint8_t> unwrap() {
```
Comment (lw) [inline tensorpipe/test/channel/channel_test.h:95]: Return by const-ref here too?

**Review by lw — APPROVED** [review]
Only a couple of minor things left, otherwise very good!

File: tensorpipe/test/channel/channel_test.h (hunk 30 above)
Comment (beauby) [reply tensorpipe/test/channel/channel_test.h:95]: Returning a const ref here is UB, as the vector is stack-allocated. Moreover, copy-elision should prevent actually copying the object.

Comment (lw) [reply tensorpipe/test/channel/channel_test.h:92]: Sorry, you're right.

**Review by lw — COMMENTED** [review]
Perfect!

**Review by lw — APPROVED** [review]
